### PR TITLE
Feature/tipsetcash

### PR DESCRIPTION
**Pull request by Elestrias**

### CODE REVIEW

File: core/miner/storage_fsm/impl/tipset_cache_impl.hpp [hunk 1]
```diff
@@ -7,20 +7,26 @@
 
 #include "miner/storage_fsm/tipset_cache.hpp"
 
+#include "api/full_node/node_api.hpp"
+
```
Comment (Alexey-N-Chernyshov) [inline core/miner/storage_fsm/impl/tipset_cache_impl.hpp:11]: Please, remove extra empty lines.

**Review by Alexey-N-Chernyshov — APPROVED** [review]

File: core/miner/storage_fsm/impl/tipset_cache_impl.hpp [hunk 2]
```diff
@@ -7,20 +7,26 @@
...
 #include "primitives/tipset/tipset_key.hpp"
 
+#include "vm/actor/builtin/types/miner/deadline.hpp"
+
 #include <shared_mutex>
 
 namespace fc::mining {
+  using api::FullNodeApi;
+  using fc::vm::actor::builtin::types::miner::QuantSpec;
```
Comment (ortyomka) [inline core/miner/storage_fsm/impl/tipset_cache_impl.hpp:20]: Remove

File: core/miner/storage_fsm/impl/tipset_cache_impl.hpp [hunk 3]
```diff
@@ -7,20 +7,26 @@
 
 #include "miner/storage_fsm/tipset_cache.hpp"
 
+#include "api/full_node/node_api.hpp"
+
 #include "primitives/tipset/tipset_key.hpp"
 
+#include "vm/actor/builtin/types/miner/deadline.hpp"
```
Comment (ortyomka) [inline core/miner/storage_fsm/impl/tipset_cache_impl.hpp:14]: Remove

File: core/miner/storage_fsm/impl/tipset_cache_impl.hpp [hunk 4]
```diff
@@ -7,20 +7,26 @@
...
+
 #include <shared_mutex>
 
 namespace fc::mining {
+  using api::FullNodeApi;
+  using fc::vm::actor::builtin::types::miner::QuantSpec;
+  using primitives::tipset::TipsetKey;
```
Comment (ortyomka) [inline core/miner/storage_fsm/impl/tipset_cache_impl.hpp:21]: Remove

File: test/core/miner/CMakeLists.txt [hunk 5]
```diff
@@ -35,6 +35,7 @@ addtest(tipset_cache_test
     )
 target_link_libraries(tipset_cache_test
     tipset_cache
+    miner_types
```
Comment (ortyomka) [inline test/core/miner/CMakeLists.txt:38]: Does it need this?

Comment (Elestrias) [reply test/core/miner/CMakeLists.txt:38]: Yes, because of deadlines

Comment (Elestrias) [reply test/core/miner/CMakeLists.txt:38]: do you mean move it to tipset_cahce.hpp ?

File: core/miner/storage_fsm/impl/events_impl.cpp [hunk 6]
```diff
@@ -47,14 +47,9 @@ namespace fc::mining {
                                             EpochDuration confidence,
                                             ChainEpoch height) {
     std::unique_lock<std::mutex> lock(mutex_);
-    auto best_tipset = tipset_cache_->best();
+    OUTCOME_TRY(best_tipset, tipset_cache_->best());
 
-    // TODO(ortyomka): [FIL-370] should be updated after update TipSetCache
-    if (!best_tipset) {
-      return EventsError::kNotFoundTipset;
```
Comment (ortyomka) [inline core/miner/storage_fsm/impl/events_impl.cpp:54]: Remove this enum class. It doesn't need anymore.

**Review by ortyomka — APPROVED** [review]

File: core/miner/storage_fsm/impl/events_impl.cpp [hunk 7]
```diff
@@ -64,13 +59,9 @@ namespace fc::mining {
       OUTCOME_TRY(handler(tipset, best_height));
 
       lock.lock();
-      best_tipset = tipset_cache_->best();
+      OUTCOME_TRY(best_tipset, tipset_cache_->best());
```
Comment (turuslan) [inline core/miner/storage_fsm/impl/events_impl.cpp:62]: duplicate/shadowed variable `best_tipset`, is it intended?  
may use `OUTCOME_TRYA(best_tipset, ...)` to assign to existing variable.

**Review by turuslan — APPROVED** [review]

**Review by Alexey-N-Chernyshov — APPROVED** [review]

**Review by Alexey-N-Chernyshov — COMMENTED** [review]
Please, check build status.

**Pull request merged**